lib/logdna/client.rb

```diff
@@ -44,95 +63,134 @@ def process_message(msg, opts = {})
 
-    def schedule_flush
-      start_timer = lambda {
-        sleep(@exception_flag ? @retry_timeout : @flush_interval)
-        flush if @flush_scheduled
-      }
-      Thread.new { start_timer.call }
+    def write_to_buffer(msg, opts)
+      Concurrent::Future.execute({ executor: @work_thread_pool }) { write_to_buffer_sync(msg, opts) }
     end
 
-    def write_to_buffer(msg, opts)
-      if @lock.try_lock
-        processed_message = process_message(msg, opts)
-        new_message_size = processed_message.to_s.bytesize
-        @buffer.push(processed_message)
-        @buffer_byte_size += new_message_size
-        @flush_scheduled = true
-        @lock.unlock
-
-        if @flush_limit <= @buffer_byte_size
+    def write_to_buffer_sync(msg, opts)
+      processed_message = process_message(msg, opts)
+      message_size = processed_message.to_s.bytesize
+
+      @lock.synchronize do
+        total_size = message_size
+        if @buffer.any?
+          total_size += @buffer[-1].size
+        end
+        @buffer.push(Message.new(processed_message, total_size))
+
+        if total_size >= @flush_threshold
+          unschedule_flush
           flush
         else
           schedule_flush
         end
-      else
-        @side_message_lock.synchronize do
-          @side_messages.push(process_message(msg, opts))
-        end
       end
     end
 
-    # This method has to be called with @lock
-    def send_request
-      @side_message_lock.synchronize do
-        @buffer.concat(@side_messages)
-        @side_messages.clear
+    ##
+    # Flushes all logs to LogDNA asynchronously
+    def flush(options = {})
+      Concurrent::Future.execute({ executor: @work_thread_pool }) { flush_sync(options) }
+    end
+
+    ##
+    # Flushes all logs to LogDNA synchronously
+    def flush_sync(options = {})
+      slices = @lock.synchronize do
+        # Slice the buffer into chunks that try to be no larger than @flush_limit. Slice points are found with
+        # a binary search thanks to the structure of @buffer. We are working backwards because it's cheaper to 
+        # remove from the tail of an array instead of the head
+        slices = []
+        while @buffer.length != 0 do
+          search_size = @buffer[-1].size - @flush_limit
+          if search_size < 0
+            search_size = 0
+          end
+
+          slice_index = @buffer.bsearch_index { |message| message.size >= search_size }
+          slices.push(@buffer.pop(@buffer.length - slice_index).map { |message| message.source })
+        end
+        slices
+      end
+
+      # Remember the chunks are in reverse order, this un-reverses them
+      slices.reverse_each do |slice|
+        if options[:block_on_requests]
+          send_request(slice)
+        else
+          Concurrent::Future.execute({ executor: @request_thread_pool }) { send_request(slice) }
+        end
       end
+    end
 
+    def send_request(slice)
+      # TODO: Remove instance-global request object
```

Comment:
is this for future todo-ing?

Reply:
Yup, leaving this here to tackle in a different PR.